File: packages/oxlint-plugin-react-doctor/src/plugin/rules/r3f/three-webgpu-no-high-precision-instancing.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getRangeStart } from "../../utils/get-range-start.js";
import { getStaticPropertyName } from "../../utils/get-static-property-name.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { readStaticBoolean } from "../../utils/read-static-boolean.js";
import { resolveExpressionKey } from "../../utils/resolve-expression-key.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { stripParenExpression } from "../../utils/strip-paren-expression.js";
import { getThreeConstructorName } from "./utils/get-three-constructor-name.js";

interface HighPrecisionAssignment {
  readonly node: EsTreeNodeOfType<"AssignmentExpression">;
  readonly rendererKey: string;
}

interface IncompatibleSceneRoot {
  readonly node: EsTreeNode;
  readonly rootKey: string;
}

interface RendererSceneRender {
  readonly node: EsTreeNodeOfType<"CallExpression">;
  readonly rendererKey: string;
  readonly rootKey: string;
}

const HIGH_PRECISION_INCOMPATIBLE_OBJECT_NAMES: ReadonlySet<string> = new Set([
  "InstancedMesh",
  "SkinnedMesh",
]);

export const threeWebgpuNoHighPrecisionInstancing = defineRule({
  id: "three-webgpu-no-high-precision-instancing",
  title: "WebGPU high precision conflicts with instanced meshes",
  category: "Correctness",
  requires: ["three:181"],
  severity: "error",
  recommendation:
    "Disable renderer.highPrecision when rendering InstancedMesh or SkinnedMesh objects",
  create: (context: RuleContext) => {
    const assignments: HighPrecisionAssignment[] = [];
    const incompatibleRoots: IncompatibleSceneRoot[] = [];
    const renders: RendererSceneRender[] = [];
    return {
      AssignmentExpression(node: EsTreeNodeOfType<"AssignmentExpression">) {
        const target = stripParenExpression(node.left);
        if (
          node.operator !== "=" ||
          readStaticBoolean(node.right) !== true ||
          !isNodeOfType(target, "MemberExpression") ||
          getStaticPropertyName(target) !== "highPrecision" ||
          getThreeConstructorName(target.object, context.scopes) !== "WebGPURenderer"
        ) {
          return;
        }
        const rendererKey = resolveExpressionKey(target.object, context);
        if (rendererKey) assignments.push({ node, rendererKey });
      },
      CallExpression(node: EsTreeNodeOfType<"CallExpression">) {
        const callee = stripParenExpression(node.callee);
        if (!isNodeOfType(callee, "MemberExpression")) return;
        const methodName = getStaticPropertyName(callee);
        if (
          methodName === "add" &&
          getThreeConstructorName(callee.object, context.scopes) === "Scene"
        ) {
          const rootKey = resolveExpressionKey(callee.object, context);
          if (!rootKey) return;
          for (const argument of node.arguments) {
            if (
              !isNodeOfType(argument, "SpreadElement") &&
              HIGH_PRECISION_INCOMPATIBLE_OBJECT_NAMES.has(
                getThreeConstructorName(argument, context.scopes) ?? "",
              )
            ) {
              incompatibleRoots.push({ node: argument, rootKey });
            }
          }
          return;
        }
        if (
          methodName !== "render" ||
          getThreeConstructorName(callee.object, context.scopes) !== "WebGPURenderer"
        ) {
          return;
        }
        const renderedRoot = node.arguments[0];
        if (!renderedRoot || isNodeOfType(renderedRoot, "SpreadElement")) return;
        const rendererKey = resolveExpressionKey(callee.object, context);
        const rootKey = resolveExpressionKey(renderedRoot, context);
        if (!rendererKey || !rootKey) return;
        renders.push({ node, rendererKey, rootKey });
        if (
          HIGH_PRECISION_INCOMPATIBLE_OBJECT_NAMES.has(
            getThreeConstructorName(renderedRoot, context.scopes) ?? "",
          )
        ) {
          incompatibleRoots.push({ node: renderedRoot, rootKey });
        }
      },
      "Program:exit"() {
        for (const assignment of assignments) {
          const assignmentStart = getRangeStart(assignment.node);
          const incompatibleRender = renders.find((render) => {
            const renderStart = getRangeStart(render.node);
            return Boolean(
              render.rendererKey === assignment.rendererKey &&
              assignmentStart !== null &&
              renderStart !== null &&
              assignmentStart < renderStart &&
              incompatibleRoots.some((root) => root.rootKey === render.rootKey),
            );
          });
          if (!incompatibleRender) continue;
          context.report({
            node: assignment.node,
            message:
              "WebGPURenderer highPrecision uses CPU 64-bit matrices that Three.js does not support with InstancedMesh or SkinnedMesh objects rendered by this renderer",
          });
        }
      },
    };
  },
});

File: native/oxlint/rules/three-webgpu-no-high-precision-instancing.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;
use oxc_syntax::operator::AssignmentOperator;
use rustc_hash::FxHashSet;

use crate::{context::LintContext, rule::Rule};

const INCOMPATIBLE_OBJECT_NAMES: [&str; 2] = ["InstancedMesh", "SkinnedMesh"];
const MESSAGE: &str = "WebGPURenderer highPrecision uses CPU 64-bit matrices that Three.js does not support with InstancedMesh or SkinnedMesh objects rendered by this renderer";

#[derive(Debug, Default, Clone)]
pub struct ThreeWebgpuNoHighPrecisionInstancing;

declare_oxc_lint!(
    /// Disallow WebGPU high precision with instanced or skinned meshes.
    ThreeWebgpuNoHighPrecisionInstancing,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Disallow incompatible WebGPU high-precision rendering.",
);

impl Rule for ThreeWebgpuNoHighPrecisionInstancing {
    fn run_once(&self, ctx: &LintContext<'_>) {
        if !has_capability_or_unspecified(ctx, "three:181") {
            return;
        }
        let mut assignments = Vec::new();
        let mut incompatible_root_symbol_ids = FxHashSet::default();
        let mut renders = Vec::new();
        for node in ctx.nodes().iter() {
            match node.kind() {
                AstKind::AssignmentExpression(assignment)
                    if assignment.operator == AssignmentOperator::Assign =>
                {
                    let Some(member_expression) = assignment.left.as_member_expression() else {
                        continue;
                    };
                    if member_expression.static_property_name() != Some("highPrecision")
                        || !matches!(
                            assignment.right.get_inner_expression(),
                            oxc_ast::ast::Expression::BooleanLiteral(value) if value.value
                        )
                        || three_constructor_api_name(member_expression.object(), ctx).as_deref()
                            != Some("WebGPURenderer")
                    {
                        continue;
                    }
                    let Some(renderer_symbol_id) =
                        resolve_stable_identifier_symbol(member_expression.object(), ctx)
                    else {
                        continue;
                    };
                    assignments.push((renderer_symbol_id, node.span()));
                }
                AstKind::CallExpression(call_expression) => {
                    let Some(member_expression) = call_expression.callee.as_member_expression()
                    else {
                        continue;
                    };
                    match member_expression.static_property_name() {
                        Some("add")
                            if three_constructor_api_name(member_expression.object(), ctx)
                                .as_deref()
                                == Some("Scene") =>
                        {
                            let Some(root_symbol_id) =
                                resolve_stable_identifier_symbol(member_expression.object(), ctx)
                            else {
                                continue;
                            };
                            if call_expression.arguments.iter().any(|argument| {
                                argument.as_expression().is_some_and(|expression| {
                                    three_constructor_api_name(expression, ctx).is_some_and(
                                        |constructor_name| {
                                            INCOMPATIBLE_OBJECT_NAMES
                                                .contains(&constructor_name.as_str())
                                        },
                                    )
                                })
                            }) {
                                incompatible_root_symbol_ids.insert(root_symbol_id);
                            }
                        }
                        Some("render")
                            if three_constructor_api_name(member_expression.object(), ctx)
                                .as_deref()
                                == Some("WebGPURenderer") =>
                        {
                            let Some(renderer_symbol_id) =
                                resolve_stable_identifier_symbol(member_expression.object(), ctx)
                            else {
                                continue;
                            };
                            let Some(rendered_root) = call_expression
                                .arguments
                                .first()
                                .and_then(oxc_ast::ast::Argument::as_expression)
                            else {
                                continue;
                            };
                            let Some(root_symbol_id) =
                                resolve_stable_identifier_symbol(rendered_root, ctx)
                            else {
                                continue;
                            };
                            if three_constructor_api_name(rendered_root, ctx).is_some_and(
                                |constructor_name| {
                                    INCOMPATIBLE_OBJECT_NAMES.contains(&constructor_name.as_str())
                                },
                            ) {
                                incompatible_root_symbol_ids.insert(root_symbol_id);
                            }
                            renders.push((renderer_symbol_id, root_symbol_id, node.span().start));
                        }
                        _ => {}
                    }
                }
                _ => {}
            }
        }
        for (renderer_symbol_id, assignment_span) in assignments {
            let has_incompatible_render = renders.iter().any(
                |(render_renderer_symbol_id, root_symbol_id, render_start)| {
                    *render_renderer_symbol_id == renderer_symbol_id
                        && assignment_span.start < *render_start
                        && incompatible_root_symbol_ids.contains(root_symbol_id)
                },
            );
            if has_incompatible_render {
                ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(assignment_span));
            }
        }
    }
}
